runtime/vm/VMSnapshotImpl.cpp

```diff
@@ -943,6 +998,207 @@ VMSnapshotImpl::saveJ9JavaVMStructures(void)
+			success = false;
+			goto done;
+		}
+		if (!saveAcquiredMonitor(threadCursor, &acquiredMonitorCursor, _vm->constantDynamicMutex, FALSE, FIXUPREFVM_CONSTANT_DYNAMIC_MUTEX)) { /* Java 11 */
+			success = false;
+			goto done;
+		}
+
+		/* fixup information for thread waiting to acquire a monitor. */
+		saveSystemMonitorWaiting(threadCursor);
+
+		/* Save thread's object monitors that are inflated and acquired. */
+		infoLen = vmFuncs->getOwnedObjectMonitors(threadCursor, threadCursor, NULL, 0);
+		if (-1 == infoLen) {
+			success = false;
+			goto done;
+		} else if (infoLen > 0) {
+			PORT_ACCESS_FROM_PORT(_portLibrary);
+
+			J9ObjectMonitorInfo *info = (J9ObjectMonitorInfo*) j9mem_allocate_memory(infoLen * sizeof(J9ObjectMonitorInfo), J9MEM_CATEGORY_CLASSES);
+			if (NULL == info) {
+				success = false;
+				goto done;
+			}
+
+			if (infoLen != vmFuncs->getOwnedObjectMonitors(threadCursor, threadCursor, info, infoLen)) {
+				j9mem_free_memory(info);
+				success = false;
+				goto done;
+			}
+
+			/* process object monitors */
+			for (IDATA i = 0; i < infoLen; i++) {
+				J9ObjectMonitor *objectMonitor = monitorTablePeek(_vm, info[i].object);
+
+				/* objectMonitor will be NULL if deflated. */
+				if (NULL != objectMonitor) {
+					if (!saveAcquiredMonitor(threadCursor, &acquiredMonitorCursor, objectMonitor->monitor, TRUE, (UDATA)info[i].object)) {
+						success = false;
+						j9mem_free_memory(info);
+						goto done;
+					}
+				}
+			}
+
+			j9mem_free_memory(info);	
+		}
+
+		/* Catch-all for threads that should not be restored and threads that currently
+		 * can't be restored because they are in a native frame.
+		 *
+		 * note: not a complete list
+		 */
+		UDATA notSafeToRestorePublicFlags = J9_PUBLIC_FLAGS_HALTED_AT_SAFE_POINT |
+											J9_PUBLIC_FLAGS_HALT_THREAD_INSPECTION |
+											J9_PUBLIC_FLAGS_THREAD_PARKED |
+											J9_PUBLIC_FLAGS_THREAD_TIMED;
+
+
+		/* Record application category type since omrthreads are not persisted. 
+		 * Daemon thread type is recorded in Java_java_lang_Thread_startImpl */
+		if (J9THREAD_CATEGORY_APPLICATION_THREAD == omrthread_get_category(threadCursor->osThread)
+			&& (NULL != threadCursor->threadObject) /* only persist java threads */
+			&& (0 == threadCursor->inNative) /* can't restore if the thread is in a native frame */
+			&& (J9_ARE_NO_BITS_SET(notSafeToRestorePublicFlags, threadCursor->publicFlags))
+		) {
+			threadCursor->privateFlags2 |= J9_PRIVATE_FLAGS2_SNAPSHOT_APPLICATION_THREAD;
+
+			/* TODO hack in ELS, will need a real solution in the future. 
+			 * Creating a copy of this large structure will impact footprint.
+			 */
+			memcpy(&(threadCursor->restoreEls), threadCursor->entryLocalStorage, sizeof(J9VMEntryLocalStorage));
+
+			/* Force refresh application thread's TLH. */
+			success = threadCursor->javaVM->memoryManagerFunctions->J9RefreshTLH(threadCursor);
+			if (false == success) {
```

Comment:
Is this call driven by intention to flush active TLH(s) for the thread (and have correspondent j9vmthread pointers to be set to NULL) before add it to snapshot?

Reply:
Yes. I was finding that some heap memory from running Java threads was not located in its expected address on the refresh run. Refreshing the TLH should from my understanding force the memory into a state where is can be consistently saved and restored.

Reply:
So I believe you are using wrong API then. I think using "flush TLH" instead of "refresh TLH" should solve your problem. Currently we do have call
```
GC_OMRVMThreadInterface::flushCachesForGC(MM_EnvironmentBase *env)
```
Most likely you don't want to expose `MM_EnvironmentBase` to caller.
You could create simple wrapper function to pass `j9vmthread`. You can get thread GC Environment `env` from `j9vmthread` as `j9vmthread->gcExtensions`

@amicic FYI

Reply:
okay thanks! I'll give that a try.

Reply:
Hey @dmitripivkine I tried replacing my TLH refresh call with flushCachesForGC here (https://github.com/theresa-m/openj9/commit/7f7e26247843be124be57b3cc727c872f7bf231c) but it resulted in the same segfualts I was seeing previously.